scripts/ensemble_kalman_paw_example.py

```diff
@@ -0,0 +1,251 @@
+
+right_paw_ensemble_stacks = np.concatenate((right_cam_ensemble_stacks[:,:,:2], right_cam_ensemble_stacks[:,:,:2]),2)
+scaled_right_paw_ensemble_stacks = remove_camera_means(right_paw_ensemble_stacks, means_camera)
+
+good_scaled_stacked_ensemble_preds = remove_camera_means(good_stacked_ensemble_preds[None,:,:], means_camera)[0]
+ensemble_pca, ensemble_ex_var = pca(good_scaled_stacked_ensemble_preds, 3)
+
+scaled_left_paw_ensemble_preds = remove_camera_means(left_paw_ensemble_preds[None,:,:], means_camera)[0]
+ensemble_pcs_left_paw = ensemble_pca.transform(scaled_left_paw_ensemble_preds)
+good_ensemble_pcs_left_paw = ensemble_pcs_left_paw[good_frames]
+
+scaled_right_paw_ensemble_preds = remove_camera_means(right_paw_ensemble_preds[None,:,:], means_camera)[0]
+ensemble_pcs_right_paw = ensemble_pca.transform(scaled_right_paw_ensemble_preds)
+good_ensemble_pcs_right_paw = ensemble_pcs_right_paw[good_frames]
+
+# # Kalman Filtering + Smoothing
+# $z_t = (d_t, x_t, y_t)$
+# $z_t = A z_{t-1} + e_t, e_t ~ N(0,E)$
+# $O_t = B z_t + n_t, n_t ~ N(0,D_t)$
+
+# ## Set parameters
+if paw == 'left':
+    save_keypoint_name = keypoint_names[0]
+    good_ensemble_pcs = good_ensemble_pcs_left_paw
+    ensemble_vars = left_paw_ensemble_vars
+    y = scaled_left_paw_ensemble_preds
+    ensemble_stacks = scaled_left_paw_ensemble_stacks
+else:
+    save_keypoint_name = keypoint_names[1]
+    good_ensemble_pcs = good_ensemble_pcs_right_paw
+    ensemble_vars = right_paw_ensemble_vars
+    y = scaled_right_paw_ensemble_preds
+    ensemble_stacks = scaled_right_paw_ensemble_stacks
+
+#compute center of mass
+#latent variables (observed)
+good_z_t_obs = good_ensemble_pcs #latent variables - true 3D pca
+
+##### Set values for kalman filter #####
+m0 = np.asarray([0.0, 0.0, 0.0]) # initial state: mean
+S0 =  np.asarray([[np.var(good_z_t_obs[:,0]), 0.0, 0.0], [0.0, np.var(good_z_t_obs[:,1]), 0.0], [0.0, 0.0, np.var(good_z_t_obs[:,2])]]) # diagonal: var
+
+A = np.asarray([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]) #state-transition matrix,
+d_t = good_z_t_obs[1:] - good_z_t_obs[:-1]
+Q = s*np.cov(d_t.T)
+
+#['pupil_top_r_x', 'pupil_top_r_y', 'pupil_bottom_r_x', 'pupil_bottom_r_y', 'pupil_right_r_x', 'pupil_right_r_y', 'pupil_left_r_x', 'pupil_left_r_y'] 
+C = ensemble_pca.components_.T # Measurement function is inverse transform of PCA
+R = np.eye(ensemble_pca.components_.shape[1]) # placeholder diagonal matrix for ensemble variance
+
+## Perform filtering
+#do filtering pass with time-varying ensemble variances
+print("filtering...")
+mf, Vf, S = filtering_pass(y, m0, S0, C, R, A, Q, ensemble_vars)
+print("done filtering")
+
+## Perform smoothing
+# Do the smoothing step
+print("smoothing...")
+ms, Vs, _ = smooth_backward(y, mf, Vf, S, A, Q, C)
+print("done smoothing")
+# Smoothed posterior over y
+y_m_smooth = np.dot(C, ms.T).T
+y_v_smooth = np.swapaxes(np.dot(C, np.dot(Vs, C.T)), 0, 1)
+
+def make_dlc_pandas_index(keypoint_names):
```

Comment:
also used in pupil script; add to other functions and import

Reply:
Not sure where to add it again

Reply:
yep sorry that wasn't clear, that's the file i meant (ensemble_kalman_filter.py)